From 1cbd1bd7ac5c00fbf79887ed2b3c94ff2c8e9897 Mon Sep 17 00:00:00 2001
Message-Id: <1cbd1bd7ac5c00fbf79887ed2b3c94ff2c8e9897.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Vidya Srinivas <vidya.srinivas@intel.com>
Date: Wed, 4 Jun 2014 09:17:26 +0530
Subject: [PATCH 089/312] FOR_UPSTREAM [VPG]: drm/i915: Writing 0 to DPLL
 creates issues.

The following changes have been ported:
	drm/i915: Modifying disable pll. Do not write 0 to DPLL register
	during disable. It creates issues during S0ix.
	(OriginalChangeID - I4b8c95b77640b5015e0b960284765e192f1ea015)

Issue: GMIN-1463
Change-Id: I949fada4d76c2a59d2e09879a2849cfe5f6dd14a
Signed-off-by: Vidya Srinivas <vidya.srinivas@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c |   13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 5ccf1d4..02ce495 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -1794,6 +1794,9 @@ static void i9xx_enable_pll(struct intel_crtc *crtc)
  */
 static void i9xx_disable_pll(struct drm_i915_private *dev_priv, enum pipe pipe)
 {
+	int reg;
+	u32 val;
+
 	/* Don't disable pipe A or pipe A PLLs if needed */
 	if (pipe == PIPE_A && (dev_priv->quirks & QUIRK_PIPEA_FORCE))
 		return;
@@ -1801,8 +1804,14 @@ static void i9xx_disable_pll(struct drm_i915_private *dev_priv, enum pipe pipe)
 	/* Make sure the pipe isn't still relying on us */
 	assert_pipe_disabled(dev_priv, pipe);
 
-	I915_WRITE(DPLL(pipe), 0);
-	POSTING_READ(DPLL(pipe));
+	reg = DPLL(pipe);
+	val = I915_READ(reg);
+	val &= ~(DPLL_VCO_ENABLE | DPLL_EXT_BUFFER_ENABLE_VLV |
+		DPLL_VGA_MODE_DIS | DPLL_INTEGRATED_CLOCK_VLV);
+	if (pipe)
+		val &= ~DPLL_REFA_CLK_ENABLE_VLV;
+	I915_WRITE(reg, val);
+	POSTING_READ(reg);
 }
 
 static void vlv_disable_pll(struct drm_i915_private *dev_priv, enum pipe pipe)
-- 
1.7.9.5

